### PR TITLE
More localization friendly source strings and context commenting

### PR DESCRIPTION
**Pull request by AmelBawa-msft**
# What was done?
- Added support for string localization with placeholders. For example:
```xml
<data name="CommandArgumentDescription" xml:space="preserve">
    <value>Filter results by command</value>
</data>
<data name="InvalidArgumentValueError" xml:space="preserve">
    <value>The value provided for the `{0}` argument is invalid; valid values are: {1}</value>
    <comment>Error message displayed when the user provides an invalid command line argument value. {0} is a placeholder replaced by the argument name. {1} is a placeholder replaced by a list of valid options.</comment>
</data>
```
```cpp
info << String::CommandArgumentDescription;                           // => Filter results by command
info << String::InvalidArgumentValueError("First"_liv, "Second"_liv); // => The value provided for the `First` argument is invalid; valid values are: Second
```
- Applied the localization placeholder change to most use cases in the code. Some remaining use cases (e.g. ShowFlow.cpp) will be handled separately as it contains localized strings with a special rendering style (e.g. text printed in bold, different color, etc...).
- Updated existing unit tests
- Documented strings with placeholders in the `winget.resw` file.

# Test
- Tested locally a randomly selected set of localized strings with placeholders.
- Added unit tests for newly added functions

### CODE REVIEW

File: src/AppInstallerCLICore/Command.cpp [hunk 1]
```diff
@@ -313,7 +280,7 @@ namespace AppInstaller::CLI
                 {
                     auto feature = ExperimentalFeature::GetFeature(command->Feature());
                     AICLI_LOG(CLI, Error, << "Trying to use command: " << *itr << " without enabling feature " << feature.JsonName());
-                    throw CommandException(Resource::String::FeatureDisabledMessage, feature.JsonName());
+                    throw CommandException(Resource::String::FeatureDisabledMessage(Utility::LocIndView{ feature.JsonName() }));
```
Comment (AmelBawa-msft) [inline src/AppInstallerCLICore/Command.cpp]: Should we change the return type of `JsonName()` to LocIndString/View instead of wrapping occurrences?
Question applies to all similar cases in this PR.

File: src/AppInstallerCLICore/Commands/FeaturesCommand.cpp [hunk 2]
```diff
@@ -53,7 +53,7 @@ namespace AppInstaller::CLI
             {
                 table.OutputLine({
                     std::string{ feature.Name() },
-                    Resource::Loader::Instance().ResolveString(ExperimentalFeature::IsEnabled(feature.GetFeature()) ? Resource::String::FeaturesEnabled : Resource::String::FeaturesDisabled),
```
Comment (AmelBawa-msft) [inline src/AppInstallerCLICore/Commands/FeaturesCommand.cpp:56]: In this PR, `Loader` class is made private to prevent directly resolving strings which could potentially result in unresolved placeholders

File: src/AppInstallerCLICore/Command.cpp [hunk 3]
```diff
@@ -114,7 +80,8 @@ namespace AppInstaller::CLI
         }
 
         // Output the command preamble and command chain
-        infoOut << Resource::String::Usage << ": winget"_liv << Utility::LocIndView{ commandChain };
+        auto commandUsage = Utility::LocIndString{ Utility::Format("{0} {1}", "winget"_lis, commandChain) };
+        infoOut << Resource::String::Usage(commandUsage);
```
Comment (JohnMcPMS) [inline src/AppInstallerCLICore/Command.cpp]: Why doesn't this just have the `{0} {1}` built in to `Usage`?

File: src/AppInstallerCLICore/Command.cpp [hunk 4]
```diff
@@ -279,10 +246,10 @@ namespace AppInstaller::CLI
         }
 
         // Finally, the link to the documentation pages
-        std::string helpLink = HelpLink();
+        auto helpLink = Utility::LocIndString{ HelpLink() };
```
Comment (JohnMcPMS) [inline src/AppInstallerCLICore/Command.cpp:249]: Make `HelpLink` return a `LocIndString(View)`.

File: src/AppInstallerCLICore/Command.cpp (hunk 1 above)
Comment (JohnMcPMS) [inline src/AppInstallerCLICore/Command.cpp]: Rather than converting these at use, make the original properly localization independent.

File: src/AppInstallerCLICore/Commands/InstallCommand.cpp [hunk 5]
```diff
@@ -124,17 +125,18 @@ namespace AppInstaller::CLI
                 std::vector<Utility::LocIndString> applicableArchitectures;
                 for (Utility::Architecture i : Utility::GetApplicableArchitectures())
                 {
-                    applicableArchitectures.emplace_back(Utility::ToString(i));
+                    applicableArchitectures.emplace_back(Utility::Format("'{0}'", Utility::ConvertFromArchitectureEnum(i)));
```
Comment (JohnMcPMS) [inline src/AppInstallerCLICore/Commands/InstallCommand.cpp]: This seems like a very heavy method for producing the string.  Are the quotes actually needed even?

File: src/AppInstallerCLICore/Commands/RootCommand.cpp [hunk 6]
```diff
@@ -178,14 +178,15 @@ namespace AppInstaller::CLI
             info << std::endl <<
                 "Windows: "_liv << Runtime::GetOSVersion() << std::endl;
 
-            info << Resource::String::SystemArchitecture << ": "_liv << Utility::ToString(Utility::GetSystemArchitecture()) << std::endl;
+            info << Resource::String::SystemArchitecture(Utility::ConvertFromArchitectureEnum(Utility::GetSystemArchitecture())) << std::endl;
 
             if (Runtime::IsRunningInPackagedContext())
             {
-                info << Resource::String::Package << ": "_liv << Runtime::GetPackageVersion() << std::endl;
+                info << Resource::String::Package(Runtime::GetPackageVersion()) << std::endl;
```
Comment (JohnMcPMS) [inline src/AppInstallerCLICore/Commands/RootCommand.cpp:186]: Even though we aren't using any emphasis here, these still fall into the "label" category and I would expect that to stay consistently separated.

File: src/AppInstallerCommonCore/Architecture.cpp [hunk 7]
```diff
@@ -187,7 +187,7 @@ namespace AppInstaller::Utility
         return Architecture::Unknown;
     }
 
-    LocIndView ToString(Architecture architecture)
+    LocIndView ConvertFromArchitectureEnum(Architecture architecture)
```
Comment (JohnMcPMS) [inline src/AppInstallerCommonCore/Architecture.cpp]: I was trying to converge around using `ToString` universally, why the change?

File: src/AppInstallerCommonCore/Public/winget/Resources.h [hunk 8]
```diff
@@ -55,5 +62,35 @@ namespace AppInstaller
...
+            ResourceOpenException(const winrt::hresult_error& hre);
+
+            const char* what() const noexcept override { return m_message.c_str(); }
+
+        private:
+            std::string m_message;
+        };
+
+        // A localized string
+        struct LocString : public Utility::LocIndString
```
Comment (JohnMcPMS) [inline src/AppInstallerCommonCore/Public/winget/Resources.h:78]: I agree that if these are defined in the same location, there does not need to be a separate type for both a localized string and a string that does not need to be localized.  Even some `LocIndString` are actually localized, they just don't need to be localized again.

`StringNotInNeedOfFurtherLocalization` is a bit verbose though...

I would consider just using `LocIndString` (and thus removing `LocString` completely) just to prevent changing every location.  Maybe the comment can be updated for now, and if we think of a better name a change that is just for the name can be done.

File: src/AppInstallerCommonCore/Public/winget/ChannelStreams.h [hunk 9]
```diff
@@ -0,0 +1,41 @@
+#pragma once
+#include "Resources.h"
+
+#define WINGET_CREATE_ISAPPROVEDFOROUTPUT_SPECIALIZATION(_t_) template <> struct IsApprovedForOutput<_t_> : std::true_type {};
+#define WINGET_CREATE_ISAPPROVEDFOROUTPUT_STRING_TYPE(_t_) " '" #_t_ "'"
+
+#define WINGET_APPROVED_FOR_OUTPUT_LIST(_f_) \
```
Comment (JohnMcPMS) [inline src/AppInstallerCommonCore/Public/winget/ChannelStreams.h:7]: Not a fan of keeping the file name now that it is here.

File: src/AppInstallerCommonCore/Public/AppInstallerStrings.h [hunk 10]
```diff
@@ -230,4 +228,31 @@ namespace AppInstaller::Utility
 
     // Converts the given hexadecimal string into bytes.
     std::vector<uint8_t> ParseFromHexString(const std::string& value, size_t byteCount = 0);
+
+    // Join a string vector using the provided separator.
+    LocIndString Join(LocIndView separator, const std::vector<LocIndString>& vector);
+
+    // Superset of std::to_string supporting string convertibles as input.
+    template <typename T>
+    Utility::LocIndString ToString(T value)
```
Comment (JohnMcPMS) [inline src/AppInstallerCommonCore/Public/AppInstallerStrings.h]: This should not return a `LocIndString`.
